### PR TITLE
SQL-based Fedora object persistence (Issue 1244)

### PR DESCRIPTION
**Pull request by seth-shaw-unlv**
**GitHub Issue**: /Islandora-CLAW/CLAW/issues/1244

# What does this Pull Request do?

Adds options to use the MySQL or Postgresql databases for Fedora object stores instead of just the simple-file configuration (still the default)

# What's new?
* Added
    - fcrepo_persistence variable to select which persistence configuration should be used
    - variables to setup the database: fcrepo_db_name, fcrepo_db_user, and fcrepo_db_password
    - task files for MySQL and Postgresql
* Removed the repository.json config file (we now download the appropriate config based on the fcrepo_persistence variable
* Does this change require documentation to be updated? Just this README (updated)
* Does this change add any new dependencies? No.
* Does this change require any other modifications to be made to the repository
 (ie. Regeneration activity, etc.)? Only if you want to migrate your existing Fedora to a SQL-based object persistence.
* Could this change impact execution of existing code? No.

# How should this be tested?

- Update your claw-playbook requirements.yml:
```
- src: https://github.com/seth-shaw-unlv/ansible-role-fcrepo
  name: Islandora-Devops.fcrepo
  version: issue-1244
```
- Vagrant up

That should give you the file-based object persistence as before. The README gives information about using the role with a SQL-based object persistence. 

**Alternatively**, you can use /Islandora-Devops/claw-playbook/pull/132 to test provisioning with a MySQL store.

# Additional Notes:

I've tested simple-file and MySQL; not with Postgresql.

# Interested parties
@dannylamb, @whikloj, @Islandora-Devops/committers

### CODE REVIEW

**Comment by whikloj** [comment]
Okay, so... I think we could remove the `repository.json`, that was added specifically to add the `claw.cnd` file to the Fedora configuration.

The `claw.cnd` just adds a bunch of namespace prefixes.

But Fedora handles namespaces better now and the `claw.cnd` could be replaced with a `namespaces.yml` type file.
https://wiki.duraspace.org/display/FEDORA51/Best+Practices+-+RDF+Namespaces

### EXCEPT!!!!
If you have used any of those prefixes and then you are stuck with them until you do an export/import of your Fedora.

If you alter your CND files and try to restart your Fedora you will get an error and your Fedora will not re-start. 

So this would be a breaking change... configuration-wise.

**Comment by seth-shaw-unlv** [comment]
I'm not sure I understand what the problem is. I don't remove claw.cnd. It is still there as it was before. 

The only configuration change is which repository.json file is used (and the downloaded version for file-simple isn't any different than the template we had).

**Comment by seth-shaw-unlv** [comment]
@whikloj Oh! I see it now....

**Comment by seth-shaw-unlv** [comment]
Okay... I think I can fix this though by having three templates and using whichever one they had configured. This won't take long to fix.

... when I get back from my next meeting...

**Comment by whikloj** [comment]
Okay, I'll open a new ticket because we should stop using the `claw.cnd`. The new way allows you to add and change prefixes without screwing up your Fedora.

File: templates/jdbc-mysql-repository.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    "workspaces" : {
+        "predefined" : ["default"],
+        "default" : "default",
+        "allowCreation" : true,
+        "cacheSize" : 10000
+    },
+    "storage" : {
+        "persistence": {
+            "type" : "db",
+            "connectionUrl": "jdbc:mysql://${fcrepo.mysql.host:localhost}:${fcrepo.mysql.port:3306}/fcrepo?createDatabaseIfNotExist=true",
```
Comment (whikloj) [inline templates/jdbc-mysql-repository.json]: If we are going to template these files anyways, I'd say we don't set the variables on the JAVA_OPTS and instead just embed them all in here. That way the database password is not visible to anyone on the machine.
So this line becomes
```
"connectionUrl": "jdbc:mysql://{{ fcrepo_db_host }}:{{ fcrepo_db_port }}/{{ fcrepo_db_name }}?createDatabaseIfNotExist=true",
```
Both `fcrepo_db_host` and `fcrepo_db_port` do not exist yet. So we'd need some reasonable defaults defined for both MySQL and PosgreSQL.

File: templates/jdbc-mysql-repository.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            "driver" : "com.mysql.jdbc.Driver",
+            "username" : "${fcrepo.mysql.username}",
```
Comment (whikloj) [inline templates/jdbc-mysql-repository.json]: Change to 
```
"username": "{{ fcrepo_db_user }}",
```

File: templates/jdbc-mysql-repository.json [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            "password" : "${fcrepo.mysql.password}"
```
Comment (whikloj) [inline templates/jdbc-mysql-repository.json]: Change to
```
"password": "{{ fcrepo_db_password }}"
```

File: templates/jdbc-mysql-repository.json [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        },
+        "binaryStorage" : {
+            "type" : "file",
+            "directory" : "${fcrepo.binary.directory:target/binaries}",
```
Comment (whikloj) [inline templates/jdbc-mysql-repository.json]: Add new `fcrepo_binary_directory` with `target/binaries` as the default. Then change this to use that variable.
```
"directory" : "{{ fcrepo_binary_directory }}",
```

File: templates/jdbc-postgresql-repository.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    "workspaces" : {
+        "predefined" : ["default"],
+        "default" : "default",
+        "allowCreation" : true,
+        "cacheSize" : 10000
+    },
+    "storage" : {
+        "persistence": {
+            "type" : "db",
+            "connectionUrl": "jdbc:postgresql://${fcrepo.postgresql.host:localhost}:${fcrepo.postgresql.port:5432}/fcrepo",
```
Comment (whikloj) [inline templates/jdbc-postgresql-repository.json]: Same changes as in the MySQL, or you could combine the two into a single template file and change the `jdbc:mysql` to `jdbc:postgresql` depending on the database type. Your call.

File: templates/jdbc-postgresql-repository.json [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            "driver" : "org.postgresql.Driver",
```
Comment (whikloj) [inline templates/jdbc-postgresql-repository.json:14]: If you want to use a single template for both databases, you'll need to alter this driver as well.

File: README.md [hunk 7]
```diff
@@ -95,6 +107,22 @@ tomcat8_java_opts:
...
+fcrepo_db_name: fcrepo
+fcrepo_db_user: fcrepo
+fcrepo_db_host: "127.0.0.1"
+fcrepo_db_port: "3306"
+
+tomcat8_java_opts:
+- -Dfcrepo.mysql.username={{ fcrepo_db_name }}
+- -Dfcrepo.mysql.password={{ fcrepo_db_password }}
+- -Dfcrepo.mysql.host={{ fcrepo_db_host }}
+- -Dfcrepo.mysql.port={{ fcrepo_db_port }}
```
Comment (whikloj) [inline README.md]: Once we have templated the `repository.json`, this information should be removed and the associated lines in the claw-playbook PR should be removed and any needed variables overridden there.

**Review by whikloj — CHANGES_REQUESTED** [review]
Sorry for the delay @seth-shaw-unlv, some changes here that make this even more production ready (IMHO).

**Comment by seth-shaw-unlv** [comment]
@whikloj, those seem reasonable to me. I start working it.

File: templates/jdbc-postgresql-repository.json (hunk 5 above)
Comment (seth-shaw-unlv) [reply templates/jdbc-postgresql-repository.json]: I'm going to keep the logic simple (right now we just choose the file based on the file-name prefixing the repository.json with which db we are using). Using one template for both the mysql and postgres would require two tasks with when statements.

**Comment by seth-shaw-unlv** [comment]
It's nice when kicking off a rebuild corresponds with leaving for another meeting...

Anyway, I think I've touched up everything, @whikloj. Give it a spin and let me know.

**Review by dannylamb — APPROVED** [review]
I've tested this and can verify it works with mysql.  I'm inclined to merge, but will defer to @whikloj to confirm his feedback has been addressed.

**Review by whikloj — APPROVED** [review]

**Comment by whikloj** [comment]
We will need to consider how we remove people from using the `claw.cnd` file in the future (https://github.com/Islandora-CLAW/CLAW/issues/804) 

I am good with this PR, but have not tested it so I'll leave the button mashing to @dannylamb

**Comment by dannylamb** [comment]
Let the mashing of the button commence

**Pull request merged**